tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -91,32 +91,141 @@ std::vector<std::vector<Long64_t>> GetFriendEntries(const std::vector<std::pair<
 }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Get and store the names, aliases and file names of the friends of the tree.
+/// \param[in] tree The main tree whose friends to 
+Internal::FriendInfo TTreeProcessorMT::GetFriendInfo(TTree &tree)
+{
+   std::vector<Internal::NameAlias> friendNames;
+   std::vector<std::vector<std::string>> friendFileNames;
+
+   const auto friends = tree.GetListOfFriends();
+   if (!friends)
+      return Internal::FriendInfo{};
+
+   for (auto fr : *friends) {
+      const auto frTree = static_cast<TFriendElement *>(fr)->GetTree();
+
+      // Check if friend tree has an alias
+      const auto realName = frTree->GetName();
+      const auto alias = tree.GetFriendAlias(frTree);
+      if (alias) {
+         friendNames.emplace_back(std::make_pair(realName, std::string(alias)));
+      } else {
+         friendNames.emplace_back(std::make_pair(realName, ""));
+      }
+
+      // Store the file names of the friend tree
+      friendFileNames.emplace_back();
+      auto &fileNames = friendFileNames.back();
+      const bool isChain = tree.IsA() == TClassRef("TChain");
+      if (isChain) {
+         const auto frChain = static_cast<TChain *>(frTree);
+         for (auto f : *(frChain->GetListOfFiles())) {
+            fileNames.emplace_back(f->GetTitle());
+         }
+      } else {
+         const auto f = frTree->GetCurrentFile();
+         fileNames.emplace_back(f->GetName());
+      }
+   }
+
+   return Internal::FriendInfo{std::move(friendNames), std::move(friendFileNames)};
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Retrieve the name of the first TTree in the first input file, else throw.
+std::string TTreeProcessorMT::GetTreeName()
+{
+   std::string treeName;
+
+   if (fFileNames.empty())
+      throw std::runtime_error("Empty list of files and no tree name provided");
+
+   ::TDirectory::TContext ctxt(gDirectory);
+   std::unique_ptr<TFile> f(TFile::Open(fFileNames[0].c_str()));
```

Comment:
Opening a file can be an 'expensive' operation if the file is remote.  However it might be to hard to update the code to avoid 'wasting' this opening by using this file(handle) later on.

Reply:
That is a fair point, but anyway this search for the name of the tree (and consequently the opening of the file) will only happen when the user does not specify that name.